File: frontend/src/stores/loom.ts
import { defineStore } from 'pinia'
import { ref } from 'vue'
import * as loomsApi from '@/api/looms'

interface Loom {
  id: string
  owner: string
  name: string
  full_name: string
  description: string
  visibility: string
  default_stream: string
  checkpoint_count: number
  stream_count: number
  pin_count: number
  spin_count: number
  created_at: string
  updated_at: string
}

export const useLoomStore = defineStore('loom', () => {
  const loom = ref<Loom | null>(null)
  const activeStream = ref('main')
  const loading = ref(false)

  async function fetchLoom(owner: string, name: string) {
    loading.value = true
    try {
      loom.value = await loomsApi.getLoom(owner, name)
      activeStream.value = loom.value?.default_stream ?? 'main'
    } finally {
      loading.value = false
    }
  }

  function clear() {
    loom.value = null
    activeStream.value = 'main'
  }

  return { loom, activeStream, loading, fetchLoom, clear }
})

File: internal/models/loom.go
package models

type Loom struct {
	ID              string  `json:"id"`
	OwnerID         string  `json:"owner_id"`
	Name            string  `json:"name"`
	Description     string  `json:"description"`
	Visibility      string  `json:"visibility"`
	DefaultStream   string  `json:"default_stream"`
	DiskPath        string  `json:"-"`
	SizeBytes       int64   `json:"size_bytes"`
	CheckpointCount int     `json:"checkpoint_count"`
	StreamCount     int     `json:"stream_count"`
	PinCount        int     `json:"pin_count"`
	SpinCount       int     `json:"spin_count"`
	SpunFrom        *string `json:"spun_from"`
	CreatedAt       string  `json:"created_at"`
	UpdatedAt       string  `json:"updated_at"`
	SyncedAt        *string `json:"synced_at"`

	// Joined fields (not always populated)
	OwnerName string `json:"owner,omitempty"`
	FullName  string `json:"full_name,omitempty"`
}

type Collaborator struct {
	LoomID     string `json:"loom_id"`
	UserID     string `json:"user_id"`
	Permission string `json:"permission"`
	CreatedAt  string `json:"created_at"`
}

type CollaboratorWithUser struct {
	UserID      string `json:"user_id"`
	Permission  string `json:"permission"`
	CreatedAt   string `json:"created_at"`
	Username    string `json:"username"`
	DisplayName string `json:"display_name"`
	AvatarURL   string `json:"avatar_url"`
}
